### ISSUE TITLE
Separation of JPMS modules (breaking change)

### ISSUE THREAD

**Issue by mkarg** [eclipse-ee4j/jaxrs-api#935]
For version 4.x of Jakarta RESTful Web Services I'd like to propose that we split the JPMS modules, so in future only those module would get included by applications that actually are needed. This is useful wherever the footprint and bootstrap time plays a role, like in cloud and edge / embedded scenarios.

At least a separation into "client" and "container" is useful, as there certainly are client-only and container-only deployments. Looking a bit deeper it might make sense to identify more modules that could get separated, which is to be discussed.

**Comment by chkal** [comment]
I like the idea. Although I must confess that I'm not fully sure if I understand all the implications.

**Comment by jansupol** [comment]
> we would also need a "common" module that includes things required by client and server (i.e. WebApplicationException, MessageBodyReader, etc.).

The client needs to be available on the server, too, so to me, it makes sense for the server to depend on the client. The other possibility is to generate two API jars, the way the WebSocket API does, the [complete API](https://repo1.maven.org/maven2/jakarta/websocket/jakarta.websocket-api/2.0.0/) contains everything, the [client API](https://repo1.maven.org/maven2/jakarta/websocket/jakarta.websocket-client-api/2.0.0/) contains just the classes required only by the client runtime. This approach can even be used before 4.0, as it does not break the compatibility.

**Comment by mkarg** [comment]
I do not see that the server MUST depend of the client. We have many web services running that do NOT consume any other services, so why should they have access to client APIs at all? It is pretty nice to only have Server and Common API and only add a dependency on Client API *if actually needed*.

**Comment by chkal** [comment]
@mkarg Just to make sure I understand correctly. So you think we should split up the API JAR into three JARs (common, server, client)? This would be a requirement for getting different JPMS modules, correct?

**Comment by mkarg** [comment]
> @mkarg Just to make sure I understand correctly. So you think we should split up the API JAR into three JARs (common, server, client)? This would be a requirement for getting different JPMS modules, correct?

In fact, JPMS itself *demands* splitting up the JAR into several JARs (one per module).

I think that common, server, client, would be a good start, but indeed, SSE could be a valid choice for another module. But actually I could imagine that in the end we want many more modules (and: *many* modules is the idea behind JPMS in general). For example, why should an application programmer have a dependency on `RuntimeDelegate`? Or maybe MVC might like to depend on just a "low level" set of interfaces ("Servlets++"), instead of *all* of them? I think if we really look into the use cases of JAX-RS, we will find *many* modules.

Having a BOM POM ontop we could provide a still easy way to depend only on what you actually want to use.

**Comment by chkal** [comment]
I agree that splitting up into common, server and client may be a good idea. Not sure if we should split up even more (for separating SSE for example). This could be "too much".

But as mentioned above, I'm not fully sure if there are any implications I'm not aware of.
